File: 프로그래머스/4/118670. 행렬과 연산/행렬과 연산.js
// Custom Deque 구현 (head, tail 포인터를 사용하는 방식)

class Deque {
    constructor(arr) {
        const entries = arr.map((element, index) => [ index, element ]);
        this.items = new Map(entries);
        this.head = 0;
        this.tail = this.items.size;
    }
    
    get size() {
        return this.items.size;
    }
    
    get isEmpty() {
        return this.size === 0;
    }
    
    unshift(node) {
        this.head -= 1;
        this.items.set(this.head, node);
    }
    
    shift() {
        if (this.isEmpty) return null;
        const node = this.items.get(this.head);
        this.items.delete(this.head);
        this.head += 1;
        return node;
    }
    
    push(node) {
        this.items.set(this.tail, node);
        this.tail += 1;
    }
    
    pop() {
        if (this.isEmpty) return null;
        this.tail -= 1;
        const node = this.items.get(this.tail);
        this.items.delete(this.tail);
        return node;
    }
    
    get first() {
        if (this.isEmpty) return null;
        return this.items.get(this.head);
    }
    
    get last() {
        if (this.isEmpty) return null;
        return this.items.get(this.tail - 1);
    }
    
    get(index) {
        if (!this.items.has(this.head + index)) return null;
        return this.items.get(this.head + index);
    }
}



function solution(rc, operations) {
    const [ rowCount, columnCount ] = [ rc.length, rc[0].length ];
    const firstColumn = new Deque(Array.from({ length: rowCount }, (_, index) => rc[index][0]));
    const lastColumn = new Deque(Array.from({ length: rowCount }, (_, index) => rc[index][columnCount - 1]));
    const rows = new Deque(Array.from({ length: rowCount, }, (_, i) => new Deque(Array.from({ length: columnCount - 2, }, (_, j) => rc[i][j + 1]))));
    
    operations.forEach((operation) => {
        if (operation === "ShiftRow") {
            firstColumn.unshift(firstColumn.pop());
            lastColumn.unshift(lastColumn.pop());
            rows.unshift(rows.pop());
        } else {
            rows.first.unshift(firstColumn.shift());
            lastColumn.unshift(rows.first.pop());
            rows.last.push(lastColumn.pop());
            firstColumn.push(rows.last.shift());
        }
    });
    
    const result = [];
    for (let i = 0; i < rowCount; i++) {
        const first = firstColumn.get(i);
        
        const row = rows.get(i);
            
        const middle = Array.from({ length: row.size }, (_, index) => row.get(index));
        const last = lastColumn.get(i);
        const completed = [ first, ...middle, last ];
        
        result.push(completed);
    }
    
    return result;
}

File: 프로그래머스/4/118670. 행렬과 연산/행렬과 연산.py
from collections import deque
# prefix sum을 사용해도 rotateshift 연산으로 인해 너무 복잡해져버리기 때문에, matrix를 deque로 관리하기
# 첫번째, 마지막 column, 이를 제외한 rows를 따로 관리하기
# shift row 연산: O(1)
# rotate 연산: O(1)
# O((r + 1) * n) = O(rn), r <= 50,000, n <= 100,000

def solution(rc, operations):
    row_count, column_count = len(rc), len(rc[0])
    first_column = deque([ rc[i][0] for i in range(len(rc)) ])
    last_column = deque([ rc[i][column_count - 1] for i in range(len(rc))])
    rows = deque([ deque(rc[i][1: column_count - 1]) for i in range(len(rc)) ])
    for operation in operations:
        if operation == "ShiftRow":
            first_column.appendleft(first_column.pop())
            last_column.appendleft(last_column.pop())
            rows.appendleft(rows.pop())
        else:
            rows[0].appendleft(first_column.popleft())
            last_column.appendleft(rows[0].pop())
            rows[row_count - 1].append(last_column.pop())
            first_column.append(rows[row_count - 1].popleft())

    result = [ [ first_column[i], *rows[i] ,last_column[i] ]  for i in range(row_count)]
    return result
    
        